docs/debugging_flang.rst

```diff
@@ -0,0 +1,214 @@
+######################
+
+Both ``flang1`` and ``flang2`` are capable of providing a large number of diagnostic outputs. By default they are written into a file ending in ``.qdbg`` in the current working directory. Alternatively you can use debug flag ``-q,0,1`` to redirect the output to standard output. 
+
+The flags are enabled by passing ``-q`` or ``-qq`` parameters, always carrying a "flag,value" pair. On ``flang`` invocation level the flags must be specified for either ``flang1`` or ``flang2``, by specifying "H" or "M" respectively.
+
+See :doc:`flang2/coding` for details of the ``flang2`` flags meanings.
+
+Example 1: print human-readable ILM representation as read by flang2 for further processing:
+
+.. code-block:: console
+    
+    $ flang -c func.f90 -Mq,0,1 -Mq,4,1
+
+Example 2: print human-readable ILI representation as used internally by flang2:
+
+.. code-block:: console
+
+    $ flang -c hello.f90 -Mq,0,1 -Mq,10,2 -Mq,10,1
+
+In the source code these flags are represented by ``flg.dbg[i]``, where ``i`` is the flag number.
+
+The ``-qq`` parameter causes debug routines to dump human-readable views of various internal data structures. The syntax is: 
+
+::
+
+    -qq,<phase_name>,<object_name>
+
+E.g. to dump flang1 AST, one can pass ``-Hq,0,1 -Hqq,parser,ast`` to the compiler, or to dump flang1 symtab, one can pass ``-Hq,0,1 -Hqq,parser,symtab`` to the compiler.
+
+The available phase names are: init, parser, bblock, vectorize, optimize, schedule, assemble, xref, unroll
+
+The available object names are: ast, dtype, std, sstd, stdp, sym, syms, symtab, allsym, stats, area, olddtype, odtype, oldsym, osym, hsym, hsyms, common, commons, nast, stdtree, stdtrees, shape, aux
+
+**Note**: In case you are passing arguments to ``flang1`` or ``flang2`` directly, you need to skip the "H" or "M" prefix and replace commas with whitespaces. For example, the two commands should give the same debugging output:
+
+::
+
+    flang -Hq,0,1 -Hq,3,8 -Hq,47,1 -Hqq,parser,symtab example.f90
+    /usr/local/bin/flang1 -q 0 1 -q 3 8 -q 47 1 -qq parser symtab example.f90
+
+
+flang1 and flang2 command line options
+######################################
+
+As mentioned in :ref:`invocation-and-debuggers-label`, you may often need to run ``flang1`` and ``flang2`` separately in the debugger and thus provide them with all the parameters that a normal ``flang`` invocation would. Unfortunately ``flang1`` and ``flang2`` do not have ``--help`` option, so here is a list of possible options and their functionality.
+
+``flang1``:
+
++--------------+--------------------------------------------------------------------------------------------+
+| Option       | Explanation                                                                                |
++==============+============================================================================================+
+| filename     | source file name (method 1 of 2)                                                           |
```

Comment:
Would you consider styling all options in this table consistently as preformatted code? For example <code>\`\`filename\`\`</code>, <code>\`\`opt\`\`</code>, <code>\`\`x\`\`</code>, <code>\`\`qq\`\`</code>, etc.